### ISSUE TITLE
direnv runs recursively when using nixos' pkgs.buildFHSUserEnv

### ISSUE THREAD

**Issue by hgl** [direnv/direnv#992]
**Describe the bug**
When direnv runs with `use flake`, and the flake uses `pkgs.buildFHSUserEnv` with both the current shell and `runScript` being zsh, direnv run recursively.

**To Reproduce**
Make sure current shell is `zsh`
Create `.envrc`:
```
use flake
```
Create `flake.nix`
```
{
  inputs.nixpkgs.url = "nixpkgs";
  inputs.flake-utils.url = "github:numtide/flake-utils";

  outputs = { self, nixpkgs, flake-utils }:
    flake-utils.lib.eachDefaultSystem (system:
      let pkgs = nixpkgs.legacyPackages.${system};
      in {
        devShell = (pkgs.buildFHSUserEnv {
          name = "fhs-test";
          targetPkgs = pkgs: with pkgs; [ unzip ];
          multiPkgs = null;
          extraOutputsToInstall = [ "dev" ];
          profile = ''
            export hardeningDisable=all
          '';
          runScript = "zsh";
        }).env;
      });
}

```
cd zsh to this directory
**Expected behavior**
Direnev run once

**Environment**
 - OS: 22.11.20220917.0cfb3c0 (Raccoon) with unstable nixpkgs
 - Shell: zsh 5.9 (x86_64-pc-linux-gnu)
 - Direnv version: 2.32.1

**Additional context**
If I change `runScript` to `"bash"`, the direnv correctly runs only once.

Here is the debug log when I use `runScript = "zsh"` and it runs recursively
```
direnv: ([/nix/store/w5v1n1kc8yw9xy970ykvcrci3dn3vk9f-direnv-2.32.1/bin/direnv export zsh]) is taking a while to execute. Use CTRL-C to give up.
direnv: export:21:51:47 cmd_export.go:22: start                                                                                                                                                                                                                                                                              direnv: export:21:51:47 cmd_export.go:35: loading RCs
direnv: export:21:51:47 config.go:208: RCFile is blank - loadedRC is nil
direnv: export:21:51:47 cmd_export.go:43: updating RC
direnv: export:update:21:51:47 cmd_export.go:46: Determining action:
direnv: export:update:21:51:47 cmd_export.go:47: toLoad: "/home/hgl/dev/lab/direnv-fhsnix/.envrc"
direnv: export:update:21:51:47 cmd_export.go:48: loadedRC: (*cmd.RC)(nil)
direnv: export:update:21:51:47 cmd_export.go:54: no RC (implies no DIRENV_DIFF),loading
direnv: export:update:21:51:47 file_times.go:192: getLatestStat: /home/hgl/dev/lab/direnv-fhsnix/.envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: export:update:21:51:47 file_times.go:192: getLatestStat: /home/hgl/.local/share/direnv/allow/694c379f1b581a693898840c8ad1cb451d809a4e538c840b77c22129eba511e1: Lstat: 1664803332, Stat: 1664803332 -> preferring Stat
direnv: 21:51:47 file_times.go:192: getLatestStat: /home/hgl/dev/lab/direnv-fhsnix/.envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: loading ~/dev/lab/direnv-fhsnix/.envrc
direnv: using flake
direnv: 21:51:47 file_times.go:192: getLatestStat: .envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: 21:51:47 file_times.go:188: getLatestStat: /home/hgl/.config/direnv/direnvrc: Lstat: 1664803243, Stat: 1 -> preferring Lstat
direnv: 21:51:47 file_times.go:192: getLatestStat: ./flake.nix: Lstat: 1664805046, Stat: 1664805046 -> preferring Stat
direnv: 21:51:47 file_times.go:192: getLatestStat: ./flake.lock: Lstat: 1664803425, Stat: 1664803425 -> preferring Stat
direnv: nix-direnv: using cached dev shell
direnv: ([/nix/store/w5v1n1kc8yw9xy970ykvcrci3dn3vk9f-direnv-2.32.1/bin/direnv export zsh]) is taking a while to execute. Use CTRL-C to give up.
direnv: export:21:51:47 cmd_export.go:22: start                                                                                                                                                                                                                                                                              direnv: export:21:51:47 cmd_export.go:35: loading RCs
direnv: export:21:51:47 config.go:208: RCFile is blank - loadedRC is nil
direnv: export:21:51:47 cmd_export.go:43: updating RC
direnv: export:update:21:51:47 cmd_export.go:46: Determining action:
direnv: export:update:21:51:47 cmd_export.go:47: toLoad: "/home/hgl/dev/lab/direnv-fhsnix/.envrc"
direnv: export:update:21:51:47 cmd_export.go:48: loadedRC: (*cmd.RC)(nil)
direnv: export:update:21:51:47 cmd_export.go:54: no RC (implies no DIRENV_DIFF),loading
direnv: export:update:21:51:47 file_times.go:192: getLatestStat: /home/hgl/dev/lab/direnv-fhsnix/.envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: export:update:21:51:47 file_times.go:192: getLatestStat: /home/hgl/.local/share/direnv/allow/694c379f1b581a693898840c8ad1cb451d809a4e538c840b77c22129eba511e1: Lstat: 1664803332, Stat: 1664803332 -> preferring Stat
direnv: 21:51:47 file_times.go:192: getLatestStat: /home/hgl/dev/lab/direnv-fhsnix/.envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: loading ~/dev/lab/direnv-fhsnix/.envrc
direnv: using flake
direnv: 21:51:48 file_times.go:192: getLatestStat: .envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: 21:51:48 file_times.go:188: getLatestStat: /home/hgl/.config/direnv/direnvrc: Lstat: 1664803243, Stat: 1 -> preferring Lstat
direnv: 21:51:48 file_times.go:192: getLatestStat: ./flake.nix: Lstat: 1664805046, Stat: 1664805046 -> preferring Stat
direnv: 21:51:48 file_times.go:192: getLatestStat: ./flake.lock: Lstat: 1664803425, Stat: 1664803425 -> preferring Stat
direnv: nix-direnv: using cached dev shell
direnv: ([/nix/store/w5v1n1kc8yw9xy970ykvcrci3dn3vk9f-direnv-2.32.1/bin/direnv export zsh]) is taking a while to execute. Use CTRL-C to give up.
direnv: export:21:51:48 cmd_export.go:22: start                                                                                                                                                                                                                                                                              direnv: export:21:51:48 cmd_export.go:35: loading RCs
direnv: export:21:51:48 config.go:208: RCFile is blank - loadedRC is nil
```

Here is the debug log when I use `runScript = "bash"` and it runs correctly:

```
direnv: export:07:48:23 cmd_export.go:22: start
direnv: export:07:48:23 cmd_export.go:35: loading RCs
direnv: export:07:48:23 config.go:208: RCFile is blank - loadedRC is nil
direnv: export:07:48:23 cmd_export.go:43: updating RC
direnv: export:update:07:48:23 cmd_export.go:46: Determining action:
direnv: export:update:07:48:23 cmd_export.go:47: toLoad: "/home/hgl/dev/lab/direnv-fhsnix/.envrc"
direnv: export:update:07:48:23 cmd_export.go:48: loadedRC: (*cmd.RC)(nil)
direnv: export:update:07:48:23 cmd_export.go:54: no RC (implies no DIRENV_DIFF),loading
direnv: export:update:07:48:23 file_times.go:192: getLatestStat: /home/hgl/dev/lab/direnv-fhsnix/.envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: export:update:07:48:23 file_times.go:192: getLatestStat: /home/hgl/.local/share/direnv/allow/694c379f1b581a693898840c8ad1cb451d809a4e538c840b77c22129eba511e1: Lstat: 1664803332, Stat: 1664803332 -> preferring Stat
direnv: 07:48:23 file_times.go:192: getLatestStat: /home/hgl/dev/lab/direnv-fhsnix/.envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: loading ~/dev/lab/direnv-fhsnix/.envrc
direnv: using flake
direnv: 07:48:23 file_times.go:192: getLatestStat: .envrc: Lstat: 1664803324, Stat: 1664803324 -> preferring Stat
direnv: 07:48:23 file_times.go:188: getLatestStat: /home/hgl/.config/direnv/direnvrc: Lstat: 1664803243, Stat: 1 -> preferring Lstat
direnv: 07:48:23 file_times.go:192: getLatestStat: ./flake.nix: Lstat: 1664840861, Stat: 1664840861 -> preferring Stat
direnv: 07:48:24 file_times.go:192: getLatestStat: ./flake.lock: Lstat: 1664803425, Stat: 1664803425 -> preferring Stat
direnv: nix-direnv: using cached dev shell
fhs-test-chrootenv:hgl@hgl-dev:~/dev/lab/direnv-fhsnix$ direnv: ([/nix/store/w5v1n1kc8yw9xy970ykvcrci3dn3vk9f-direnv-2.32.1/bin/direnv export zsh]) is taking a while to execute. Use CTRL-C to give up.
```
Notice the last line, it was prompted differently. First the prompt `fhs-test-chrootenv:hgl@hgl-dev:~/dev/lab/direnv-fhsnix$` is displayed, then after a while, the rest line is displayed in red. And it seems odd that direnv is trying to `export zsh` in a bash.

**Comment by seanybaggins** [comment]
Same issue

**Comment by fnune** [comment]
Seeing the same problem here. I have worked around it by putting this in my `.envrc`:

```sh
if [ -z "$IN_NIX_SHELL" ]; then
  use flake
fi
```

This is only when using `runScript = "zsh"`. `runScript = "bash"` is fine/does not recurse.

Additionally, there's the problem that Direnv can't exit the subshell when leaving the directory. But that's another issue: https://github.com/direnv/direnv/issues/798

**Comment by fnune** [comment]
I believe my workaround (https://github.com/direnv/direnv/issues/992#issuecomment-1744989487) is not sufficient. I start to see this inside the shell:

```
=> direnv: ([/nix/store/3id2lsis3640a27pk8wh2z6xli88jmxk-direnv-2.32.2/bin/direnv export zsh]) is taking a while to execute. Use CTRL-C to give up.
```

This gets echoed out in red some seconds after `cd`ing into the directory. It makes me think that something's not right about my workaround.

**Comment by inmaldrerah** [comment]
For the issue starter, I think it did not run recursively with `"bash"` because direnv is only hooked to the main shell, i.e. Zsh, and not Bash.

For the workaround, calling use flake still gets into the fhs environment, which calls the inner shell. Since it is part of the direnv `.envrc` script execution, direnv will complain that it is not terminating. Also, you need to `exit` from the inner shell rather than just `cd` out.

**Comment by Oudwins** [comment]
Has there being any progress on fixing this?

**Comment by mastermach50** [comment]
> Bumping this thread

Same issue. direnv keeps running recursively when using `buildFHSEnv` @fnune 's solution works but I still get 1 extra iteration.

**Comment by Murazaki** [comment]
Can't we just check if we are already in direnv here with `DIRENV_IN_ENVRC` ?

https://github.com/direnv/direnv/blob/dfdbc97a38cb72d3cf93e22ec4b573faafef30b5/internal/cmd/shell_zsh.go#L10-L14

**Comment by Murazaki** [comment]
This completely fixed it for me :

```
if [ -z "$DIRENV_IN_ENVRC" ]; then
  use flake
fi
```

**Comment by inmaldrerah** [comment]
> This completely fixed it for me :
> 
> if [ -z "$DIRENV_IN_ENVRC" ]; then
>   use flake
> fi

Will you still be able to enter a sub-direnv in this case? E.g. if you have a nix fhs direnv in /a, and another plain nix/nix fhs direnv in /a/b.

**Comment by Murazaki** [comment]
> > This completely fixed it for me :
> > if [ -z "$DIRENV_IN_ENVRC" ]; then
> > use flake
> > fi
> 
> Will you still be able to enter a sub-direnv in this case? E.g. if you have a nix fhs direnv in /a, and another plain nix/nix fhs direnv in /a/b.

Seems like I was to hasty. The env was already set up and the direnv was not reloading at all.
There's an issue here because we are running this in the middle of a direnv command. we need to know where we are in the process, so we can run the proper action accordingly.
I'm not sure if there's a proper way to know if we are in a FHS direnv related environment and which, I'm trying with DIRENV_FILE but the same inconsistency occurs (variable is unchanged during the process).

**Comment by Murazaki** [comment]
this might work
```bash
if [ "$FHS_CURRENT" != "$1" ]; then
  export FHS_CURRENT=$1
  use flake
fi
```

**Comment by tavin** [comment]
Sadly I think the solution is to give up on this idea? You need a long-lived subshell to profit from the FHS container env right? But direnv wants the subshell to terminate almost immediately so it can scrape environment variables and copy them into the initial shell.

Sure you can hack things by adding conditionals to suppress the recursion but in the end you've made direnv essentially launch `nix-shell` or `nix develop` for you after `cd`. Now you have to `exit` and only then has direnv loaded. This finally leaves you in an ambiguous nether-shell until another `cd` makes direnv unload.

For me, dropping `.envrc` and running `nix-shell` by hand is better than this.

There is not really a bug here nor a feature that direnv can implement...

**Comment by Sewer56** [comment]
Yeah; I ran into this and it's a huge PITA.
Working with some Rust code that decided to pull OpenSSL to be able to do HTTP calls.

When I use `buildFHSEnv`, the environment never properly activates as a result of this; so `rust-analyzer` (the language server) just dies when investigating native dependencies. In the end you wind up with this weird case where going from `mkShell` to `buildFHSEnv` you broke everything in your IDE, but fixed everything in the shell.

So to make things work, you first have to load the FHS environment, then have that load the direnv (recursively). And from there you gotta open the code editor.

If your code editor uses IPC and will tell an existing process to open a new window- good luck.

----

Still not sure what the most ergonomic approach is; but @tavin above is spot on- and this is quite a PITA.

When someone else opens a repository; the expectation should somewhat be that the environment is automatically set up- but this is not quite possible here. It's unfortunate.

**Comment by ExFed** [comment]
I've been running into this same issue while using GNU Guix. I tried the above `export` trick, and also ran into the same error (which made me think something else was going to break). I was poking around in the `direnv stdlib` dump, and noticed near the top this little gem:

```bash
# This variable can be used by programs to detect when they are running inside
# of a .envrc evaluation context. It is ignored by the direnv diffing
# algorithm and so it won't be re-exported.
export DIRENV_IN_ENVRC=1
```

I figured "what could possibly go wrong?"

```bash
$ cat .envrc
if [ -z "$DIRENV_IN_ENVRC" ]; then
  exec guix time-machine -C channels.scm -- shell -m manifest.scm
fi
```

So far it seems to be working without any hiccups.
